x-pack/plugins/security/server/saved_objects/saved_objects_security_extension.ts

```diff
@@ -208,15 +663,756 @@ export class SavedObjectsSecurityExtension implements ISavedObjectsSecurityExten
+        } authorization`
+      );
+    }
+
+    const enforceMap = new Map<string, Set<string>>();
+    const spacesToAuthorize = new Set<string>([namespaceString]); // Always check authZ for the active space
+
+    for (const obj of objects) {
+      const {
+        type,
+        objectNamespace: objectNamespace,
+        existingNamespaces: existingNamespaces,
+      } = obj;
+      const objectNamespaceString = objectNamespace ?? namespaceString;
+
+      const spacesToEnforce = enforceMap.get(type) ?? new Set([namespaceString]); // Always enforce authZ for the active space
+      spacesToEnforce.add(objectNamespaceString);
+      enforceMap.set(type, spacesToEnforce);
+      spacesToAuthorize.add(objectNamespaceString);
+
+      for (const space of existingNamespaces ?? []) {
+        spacesToAuthorize.add(space); // existing namespaces are included so we can later redact if necessary
+      }
+    }
+
+    const authorizationResult = await this.authorize({
+      actions: new Set([action]),
+      types: new Set(enforceMap.keys()),
+      spaces: spacesToAuthorize,
+      enforceMap,
+      auditOptions: { objects },
+    });
+
+    return authorizationResult;
+  }
+
+  async authorizeDelete<A extends string>(
+    params: AuthorizeDeleteParams
+  ): Promise<CheckAuthorizationResult<A>> {
+    return this.internalAuthorizeDelete({
+      namespace: params.namespace,
+      // force existing namespaces to be undefined because we do not want to
+      // check them in non-bulk delete
+      objects: [{ ...params.object, existingNamespaces: undefined }],
+    });
+  }
+
+  async authorizeBulkDelete<A extends string>(
+    params: AuthorizeBulkDeleteParams
+  ): Promise<CheckAuthorizationResult<A>> {
+    return this.internalAuthorizeDelete(params, { forceBulkAction: true });
+  }
+
+  private async internalAuthorizeDelete<A extends string>(
+    params: AuthorizeBulkDeleteParams,
+    options?: InternalAuthorizeOptions
+  ): Promise<CheckAuthorizationResult<A>> {
+    const namespaceString = SavedObjectsUtils.namespaceIdToString(params.namespace);
+    const { objects } = params;
+    const enforceMap = new Map<string, Set<string>>();
+    const spacesToAuthorize = new Set<string>([namespaceString]); // Always check authZ for the active space
+
+    const action =
+      options?.forceBulkAction || objects.length > 1
+        ? SecurityAction.BULK_DELETE
+        : SecurityAction.DELETE;
+
+    this.validateObjectsArray(objects, action);
+
+    for (const obj of objects) {
+      const { type } = obj;
+      const spacesToEnforce = enforceMap.get(type) ?? new Set([namespaceString]); // Always enforce authZ for the active space
+      enforceMap.set(type, spacesToEnforce);
+      if (action === SecurityAction.BULK_DELETE && obj.existingNamespaces) {
+        for (const space of obj.existingNamespaces) {
+          spacesToAuthorize.add(space); // existing namespaces are included when performing bulk delete only
```

Comment:
I can't seem to find a reason this is different than bulkDelete other than that we just never supported deleting a single multi namespace saved object that existing in multiple spaces whereas we seem to support that for bulkDelete?

For my future self and other future readers it would be useful to explain why this is special?

Reply:
Looking through the code again, I think we can eliminate this difference. The original implementation of delete did not use existing namespaces in authorization, and though bulk delete does check them, it doesn't enforce authorization in existing spaces. They are only used for checking authorization and building the authorization map...which is also not used - we are only truly concerned with enforcing the current space.

TL;DR: I will update delete to check authz on existing spaces to keep consistency and reduce confusion.

Reply:
After looking back through the code and changes, here's the summary:
- we only really care about enforcing the current space in authorization of delete
- existing object namespaces are only considered in order to require the force flag
- (non-bulk) delete doesn't pass existing namespaces in such that it can save the repo from a query if authorization fails

I am going to reshape the parameters to the methods such that this is less confusing and add some comments to explain.